src/modules/audio/index.ts

```diff
@@ -5,4 +5,196 @@
+    _inputsList: new Array<MediaDeviceInfo>(),
+
+    _awaitingCapturePermissions: false,
+    _hasInputAccess: false,
+
+    _stream: <Nullable<MediaStream>><unknown>undefined,
+
+    _currentInputDevice: <Nullable<MediaDeviceInfo>><unknown>undefined,
+
+    setAwaitingCapturePermissions(isAwaitingCapturePermissions: boolean): void {
+        Audio._awaitingCapturePermissions = isAwaitingCapturePermissions;
+        Store.commit(Mutations.MUTATE, {
+            property: "audio.awaitingCapturePermissions",
+            value: isAwaitingCapturePermissions
+        });
+    },
+
+    setCurrentInputDevice(pDeviceInfo: MediaDeviceInfo): void {
+        Log.info(Log.types.AUDIO, `Successfully set input device ${pDeviceInfo.label}`);
+        Audio._currentInputDevice = pDeviceInfo;
+
+        Store.commit(Mutations.MUTATE, {
+            property: "audio.currentInputDevice",
+            value: pDeviceInfo
+        });
+    },
+
+    async matchDeviceWithInputsList(tracks: MediaStreamTrack[]): Promise<void> {
+        // Audio inputs do not have labels if the permission has not been received.
+        // Therefore, we update the list after success.
+        const inputsList = await Audio.requestInputsList();
+
+        for (let i = 0; i < tracks.length; i++) {
+            inputsList.forEach((device) => {
+                if (device.deviceId === tracks[i].getSettings().deviceId) {
+                    Audio.setCurrentInputDevice(device);
+                }
+            });
+        }
+    },
+
+    /**
+     * Fetch, save, and return the list of media devices of the specified kind.
+     * Fetches "audioinput" devices if no alternative specified.
+     *
+     * @returns {Promise<MediaDeviceInfo>[]} list of "audioinput" media devices
+     */
+    async requestInputsList(pKind = "audioinput"): Promise<MediaDeviceInfo[]> {
+        const devices = await navigator.mediaDevices.enumerateDevices();
+        const inputsList = devices.filter((d) => d.kind === pKind);
+        Audio.saveInputsList(inputsList);
+        return inputsList;
+    },
+
+    /**
+     * Set the list of known audio input devices. This remembers the list locally and
+     * also updates the Vuex store with the new information.
+     *
+     * @param {MediaDeviceInfo[]} pAudioInputs The list of audioinput devices to remember
+     */
+    saveInputsList(pAudioInputs: MediaDeviceInfo[]): void {
+        Audio._inputsList = pAudioInputs;
+
+        Store.commit(Mutations.MUTATE, {
+            property: "audio.inputsList",
+            value: pAudioInputs
+        });
+    },
+
+    async requestInputAccess(): Promise<Nullable<MediaStream>> {
+        if (Audio._stream) {
+            return Audio._stream;
+        }
+        const stream = await navigator.mediaDevices.getUserMedia({ audio: true, video: false });
+        if (stream) {
+            await Audio.handleRequestInputSuccess(stream);
+            return Audio._stream;
+        }
+        return null;
```

Comment:
The function returns a Promise<boolean>. The boolean is 'true' on a successful login.

Reply:
The type "Nullable<T>" is either "T | null | undefined". That is, this function returns a Promise of a value that is either a MediaStream or a 'null'. The caller checks if it gets a MediaStream back, otherwise some error has happened.